06.wc/rwc

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require 'optparse'
+require 'etc'
+require 'pathname'
+
+RENDERING_COL_LENGTH = 8
+
+def main
+  option = ARGV.getopts('l')
+
+  if ARGV[0]
+    arguments = ARGV
+    is_arguments_file = true
+  else
+    arguments = $stdin.readlines
+    is_arguments_file = false
+  end
+
+  run_wc(arguments, is_arguments_file, option)
+end
+
+def run_wc(arguments, is_arguments_file, option)
+  if is_arguments_file
+    results = retrieve_files(arguments)
+    render_argument_files(results, option)
+  else
+    results = build_data(arguments)
+    render_standard_input(results, option)
+  end
+end
+
+def build_data(arguments)
+  {
+    lines: count_number_of_lines(arguments),
+    words: count_number_of_words(arguments),
+    bytes: count_number_of_bytes(arguments)
+  }
+end
+
+def count_number_of_lines(arguments)
+  # mac の wc に準拠して、lines ではなく改行コードの数を count
+  arguments.instance_of?(Array) ? arguments.length : arguments.count("\n")
+end
+
+def count_number_of_words(arguments)
+  if arguments.instance_of?(Array)
+    arguments.sum do |argument|
+      argument.split.length
+    end
+  else
+    arguments.strip.split(/\s+/).length
+  end
+end
+
+def count_number_of_bytes(arguments)
+  arguments.instance_of?(Array) ? arguments.sum(&:bytesize) : arguments.bytesize
+end
+
+def retrieve_files(arguments)
+  arguments.map do |argument|
+    file_path = Pathname("./#{argument}")
+
+    File.exist?(file_path) ? get_file_data(file_path) : "wc: #{argument}: open: No such file or directory"
+  end
+end
+
+def get_file_data(file_path)
+  stat = File.lstat(file_path)
+
+  if stat.directory?
+    "wc: #{File.basename(file_path)}: read: Is a directory"
+  else
+    file_string = File.read(file_path)
+    result = build_data(file_string)
+    result[:file_name] = File.basename(file_path)
+    result
+  end
+end
+
+def render_standard_input(result, option)
+  puts justify_output(result, option)
+end
+
+def render_argument_files(results, option)
+  results.each do |result|
+    if result.instance_of?(Hash)
+      puts justify_output(result, option, result[:file_name])
+    else
+      puts result
+    end
+  end
+
+  (puts justify_output(sum_file_size(results), option, 'total')) if results.length > 1
```

Comment:
この部分、僕は以下のようにした方が可読性が高いと考えたのですが、それだと Rubocop で`rwc:99:3: C: Style/GuardClause: Use a guard clause (return unless results.length > 1) instead of wrapping the code inside a conditional expression.`と警告が出たので上記のように書いています。

```ruby
if results.length > 1
  total = sum_file_size(results)
  puts justify_output(total, option, 'total')
end
```

```ruby
total = sum_file_size(results)
(puts total, option, 'total')) if results.length > 1
```
という記述も考えたのですが、それだと利用しない場合がある変数が存在してしまうので採用しませんでした。

Reply:
> 以下のようにした方が可読性が高いと考えた

僕の感覚でもそう感じます。そういう場合は理由を添えた上で該当箇所だけ、引っかかっている指摘を無効化してもよいかなと思います。対応するかどうかはおまかせします

参考
https://docs.rubocop.org/rubocop/1.0/configuration.html#disabling-cops-within-source-code